en/lessons/basics/collections.md

```diff
@@ -3,209 +3,774 @@ version: 1.3.1
+From its definition we can see that `NonEmpty` requires the first element to be
+present. Despite its similarity to the sum constructor `|`, the `:|` constructor
+of `NonEmpty` is just a plane old binary constructor; just a fancy looking
+tuple!
+
+This handles the partiality problem the exact opposite way as the
+`Maybe` solution. Instead of forcing the caller of the function to guard against
+the ill-defined case when handling the result of the function, it forces them to
+construct a valid input up front (when calling the function).
+
+```haskell
+ghci> :t Data.List.NonEmpty.head
+Data.List.NonEmpty.head :: NonEmpty a -> a
+ghci> Data.List.NonEmpty.head (1 :| [])
+1
+ghci> Data.List.NonEmpty.head []
+
+<interactive>:28:25: error:
+    • Couldn't match expected type: NonEmpty a
+                  with actual type: [a0]
+    ...
+```
 
-```elixir
-iex> [head | tail] = [3.14, :pie, "Apple"]
-[3.14, :pie, "Apple"]
-iex> head
-3.14
-iex> tail
-[:pie, "Apple"]
+Notice that this time the error is not a runtime exception but a type error, the
+compiler is politely telling us that we tried to use a (potentially empty) list
+rather than the required `NonEmpty` type.
+
+## Assoc lists
+
+So far we have only been able to access values via position indexed lookup, or
+pattern matching. However, one of the most common use cases for containers is
+acting as a key value store. Assoc(iation) lists provide this by combining
+2-tuples and regular lists. Since assoc lists are really just the combination
+of two existing data types, the only thing we need is a lookup function, which
+is provided in the `Data.List` module in base.
+
+```haskell
+ghci> assoc = [("foo", True), ("bar", False)]
+ghci> :t assoc
+assoc :: [(String, Bool)]
+ghci> :t lookup
+lookup :: Eq a => a -> [(a, b)] -> Maybe b
+ghci> lookup "foo" assoc
+Just True
+ghci> lookup "bar" assoc
+Just False
+ghci> lookup "baz" assoc
+Nothing
 ```
 
-## Tuples
+We can see the pattern, again, where undefined behaviour is encoded using the
+`Maybe` type. This allows us to have complete functions with errors at compile
+time rather than runtime.
+
+It is also interesting to note the `Eq a` constraint on the "key" which allows
+the lookup function to do an equality comparison on them.
+
+While assoc lists are a nice introduction to key value collections that build
+on the previous types we learned about, they are not particularly useful. A list
+simply isn't a very good data structure for lookup, as it provides worst case
+`O(n)` asymptotics. Assoc lists are usually an intermediate data structure
+which Haskell programmers will usually convert into a `Map`. Although this
+conversion is itself an `O(n*log n)` operation `Map` provides an `O(log n)`
+lookup, so the conversion cost is ammortized very quickly.
+
+## Sets
```

Comment:
This part is very complete but I fear that we are putting too much out there!  
Could you please take two, three sentences for each "major" operation? I think the bulk of the content of this section is more appropriate for a post in the "blog" section of the website (where we can be much more prolific)

Reply:
You're breaking my heart! But I think I agree, I will try and save the content for a blog post, and shorten this section.